### ISSUE TITLE
Example notebook for converting panel data to required data format[DOC]

### ISSUE THREAD

**Issue by bendykstra94** [alan-turing-institute/sktime#543]
Hi there! Great project and work that you all have been doing :)

As a new comer to using this package, I've been trying to figure out how to format the data that I have so that it can be used by all of the awesome methods that you guys have made. I've taken a look at [load data](https://github.com/alan-turing-institute/sktime/blob/master/examples/loading_data.ipynb) example where there is extensive writing about loading from a .ts file as well as how to use long format data with sktime. I've also looked at the other examples but they all have the data in it's correct format

It says the from_long_to_nested method converts rows from a long-table schema data frame _assuming_ each row contains information for `case_id, dimension_id, reading_id, value`. From my understanding (please correct me if i'm wrong) long format data should be in the format (cases, feature dim, time dim). For example, if I have 5 customers, 3 different columns for each one, and 6 time observations for each customer, the dimensions would 5 x 3 x 6. 

If we begin with this data format (I feel like this is a pretty common initial representation):

| CustomerId      | date | feature1 | feature2 | feature3 | class
| ----------- | ----------- | ----------- | ----------- | ----------- |----------- |
| 1      | 2010-10-15 | 1 | 2| 3| 'a'|
| 1      | 2010-10-16 | 3 | 4| 5| 'b'|
.
.
.
| 2   | 2010-10-15 | 1 | 2| 3| 'c'|
| 2      | 2010-10-16 | 3 | 4| 5|'d'|
.
.

How do the above columns map to the `case_id, dimension_id, reading_id, value` representation?
CustomerId -> case_id, correct?

Could there be an example that uses some concrete examples of the mapping?

Could there also be an example that shows how to prepare this data for use by `from_long_to_nested`? 

Even better might be a method `df_to_long(case_id_identifier, feature_name_list, date_column_identifier)`  that converts a dataframe that is [tidy](https://www.ssc.wisc.edu/sscc/pubs/DWE/book/6-2-tidy-data.html) to the required long format.

case_id_identifier would be the unique identifier (CustomerId in my case), feature_name_list an array of strings that identify the columns you want to use, and date_column_identifier a string that identifies the date column

Thanks so much!

**Comment by bendykstra94** [comment]
Looks like this might have been asked already https://github.com/alan-turing-institute/sktime/issues/265

**Comment by mloning** [comment]
Hi @bendykstra94 thanks for creating this issue! 

I think one problem is that there isn't really a consensus representation for panel data.  

What you call `features` is what is called `dimensions` in that notebook. So "melting" the feature columns into a feature name and value column should do the trick.  

We'd appreciate a PR to add those discussion to the data loading notebook!

**Comment by bendykstra94** [comment]
Hi @mloning thanks for your answer!

I guess a lot of the confusion for me stems from the lack of documentation and verbiage around the format of the example data that any of the models take.

In the motions data set, 

> Participants were required to record motion a total of five times, and the data is sampled once every tenth of a second, for a ten second period.

What exactly does a row in this data set represent? How many observations per a label are there? What is the definition of a case in this dataset?

For instance, what is a `case`? In my (edited) example above, is it one row where I have a label for every observation?


I agree that there is not a consensus about how to represent panel data. However, this package seems to be pretty opinionated about what the format of the input data should be, but there is no documentation or examples about what that input format should be. Can I use a numpy array of shape `(cases x n dimensions x observations per case)` as an input? 

Sorry for all of my questions, I come from a cross sectional background where the formatting is more straight forward :) I hope that in answering some of these questions, others like me will be able to use and improve this package.

Once I understand what's going on, I'd like to add a PR

**Comment by mloning** [comment]
Yes a numpy array with `(cases x n dimensions x observations per case)` works! 

A case is essentially a statistical idea, it's an experimental unit, sometimes it's called samples (scikit-learn) or examples or instances. Our preference is to use instances as samples can be ambiguous with regard to time series observations.

The other currently supported data format is a nested `pd.DataFrame`:

```python
from sktime.datasets import load_arrow_head
X, y = load_arrow_head(return_X_y=True)  # X is a nested pd.DataFrame
```

Part of the problem is that there isn't an equivalent to numpy for panel data or even multivariate time series data. I agree we need to add more documentation on the data formats! Your help would be extremely welcome here since you seem to have thought about data formatting already.

**Comment by fkiraly** [comment]
Actually, thinking about long format as an input/output format, we should maybe support does name a lot of sense - we've been discussing around the merits of nested vs array, with nested being "abuse of API" while array is common but does not allow different length sequences.

Data frame in long format (which we may not have given enough focus in our discussions) is both common in practice, and allows for different length sequences!
Though it might require some specific column naming convention to make this work.

**Comment by bendykstra94** [comment]
@fkiraly different sequence length is something that's been dealt with in other packages, have a look at how Keras pads time sequences of different lengths https://keras.io/api/preprocessing/timeseries/

give this a look?
```
def tidy_to_long(df, case_id_identifier, feature_arr, n_observations_per_label):
    n_features = len(feature_arr)
    n_cases = len(df[case_id_identifier].unique())
    
    rows_per_case = n_observations_per_label * n_features
    
    total_rows = int(n_cases * n_observations_per_label * n_features)
    
    case_ids = np.empty(total_rows, dtype=np.int)
    idxs = np.empty(total_rows, dtype=np.int)
    dims = np.empty(total_rows, dtype=np.int)
    
    for i in range(total_rows):
        case_ids[i] = int(i / rows_per_case)
        rem = i % rows_per_case
        dims[i] = int(rem / n_observations_per_label)
        idxs[i] = rem % n_observations_per_label
    
    rows = range(len(df[feature_arr]))
    dim_ids = range(n_features)
    
    # not sure if the dtypes are preserved here
    vals = df[feature_arr].values.reshape((n_features*len(df)) )
    
    new_df = pd.DataFrame()
    new_df["case_id"] = pd.Series(case_ids)
    new_df["dim_id"] = pd.Series(dims)
    new_df["reading_id"] = pd.Series(idxs)
    new_df["value"] = pd.Series(vals)
    
    return new_df
                
```

Given this tidy data
```
data = {
    'personId': [1,1,2,2,3,3,4,4,5,5,6,6],
    'date': ['2010-10-15', '2010-10-16','2010-10-15', '2010-10-16','2010-10-15',
             '2010-10-16','2010-10-15', '2010-10-16','2010-10-15', '2010-10-16','2010-10-15', '2010-10-16'],
    'age': [30,30,35,35,32,32,21,21,25,25,18,18],
    'height': [67,67,69,69,72,72,73,73,63,63,59,59],
    'bodyTemp': [97.6,97.8, 96.9,96.8,98.1,98.5,100.0,98.8,97.6, 96.8,97.2,98.0],
    'label': [None,1,None,0,None,1,None,1,None,0,None,1]
}

# 6 cases, 3 dimensions, 2 observations per label
tidy = pd.DataFrame(data)
```

we can call
```
long_format = tidy_to_long(df = tidy,
  case_id_identifier = 'personId', 
  feature_arr = ['age', 'height', 'bodyTemp'],
  n_observations_per_lable = 2
)

long_format.head()
```
and get

case_id | dim_id | reading_id | value
-- | -- | -- | --
0 | 0 | 0 | 30.0
0 | 1 | 0 | 67.0
0 | 2 | 0 | 97.6
1 | 0 | 0 | 30.0
1 | 1 | 0 | 67.0
1 | 2 | 0 | 97.8
2 | 0 | 0 | 35.0
2 | 1 | 0 | 69.0
2 | 2 | 0 | 96.9
-- | -- | -- | -- | --



and then use `from_long_to_nested`

**Comment by fkiraly** [comment]
@bendykstra94, thanks for the pointers - what I was trying to say though is that it might be worth considering the long format as a native input/output format for the panel setting. The"typical" conversion long/wide I've seen before.

**Comment by RNKuhns** [comment]
The loading_data.ipynb in examples has been updated based on the work in pull request #553. There is now additional functions to convert between a variety of data formats and the nested format currently used by sktime.

**Comment by TonyBagnall** [comment]
hi, I think this has now been addressed, if there are issues with the current state feel free to open a new one
